test/prober.go

```diff
@@ -0,0 +1,266 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// route.go provides methods to perform actions on the route resource.
+
+package test
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+	"testing"
+	"time"
+
+	pkgTest "github.com/knative/pkg/test"
+	"github.com/knative/pkg/test/logging"
+	"github.com/knative/pkg/test/spoof"
+	"golang.org/x/sync/errgroup"
+)
+
+// Prober is the interface for a prober, which checks the result of the probes when stopped.
+type Prober interface {
+	// SLI returns the "service level indicator" for the prober, which is the observed
+	// success rate of the probes.  This will panic if the prober has not been stopped.
+	SLI() (total int64, failures int64)
+
+	// Stop terminates the prober, returning any observed errors.
+	// Implementations may choose to put additional requirements on
+	// the prober, which may cause this to block (e.g. a minimum number
+	// of probes to achieve a population suitable for SLI measurement).
+	Stop() error
+}
+
+type prober struct {
+	// These shouldn't change after creation
+	logger        *logging.BaseLogger
+	domain        string
+	minimumProbes int64
+
+	// m guards access to these fields
+	m        sync.Mutex
+	requests int64
+	failures int64
+	stopped  bool
+	err      error
+}
+
+// prober implements Prober
+var _ Prober = (*prober)(nil)
+
+// SLI implements Prober
+func (p *prober) SLI() (int64, int64) {
+	p.m.Lock()
```

Comment:
Lock -> RWLock and here RLock()/RUnlock()?

Reply:
Sure.